src/ripple/nodestore/impl/DatabaseShardImp.cpp

```diff
@@ -733,162 +765,215 @@ DatabaseShardImp::onChildrenStopped()
 void
 DatabaseShardImp::import(Database& source)
 {
+    assert(init_);
```

Comment:
I removed the lock to simplify things when invoking other `DatabaseShardImp` functions from inside of `import`. Apart from unit tests, `import` is invoked only while the application is still initializing and before the `JobQueue` has been started.

If you're fine with this, perhaps you'd suggest that this design choice warrants a comment?

Reply:
Yup, I'm fine with this but, as you say, at the minimum a comment is called for.

Beyond the comment, I think that it's probably important to add some code that verifies your invariants, since your invariants are really unusual.  Maybe you can check at the beginning and end that the `JobQueue` is not running?  Are there other unusual conditions your approach is dependent on?  If so, can you find a way to verify that the conditions are met?

I'm not saying that you haven't coded up all of these preconditions just fine today.  But I'm suggesting that unusual conditions like these are fragile and tend to break over time.  So you should add some checks that the stuff you depend on is still in place in the future.

Reply:
Done. Added a comment to explain the missing lock and a check to make sure the app hasn't yet started.

Reply:
> @undertome What do you mean when you say "before the `JobQueue` has been started"? `JobQueue` does not have a `start` method, or either of the start callbacks in the `Stoppable` framework, `onPrepare` and `onStart`. Are you referring to the call to `JobQueue::setThreadCount` in `ApplicationImp::setup`?

Well, before the app has started via the `RootStoppable` interface.

Reply:
> ::import needs to work asynchronously, it cannot block. A full history server would not be able to start for months as this works now. It's a good job for the Taskqueue. You'll need to increase the worker thread count to 2 and likely reduce that to 1 upon import completion.

Done